File: src/org/elixir_lang/GenericServer.java
package org.elixir_lang;

import com.ericsson.otp.erlang.*;
import org.jetbrains.annotations.NotNull;

/**
 * Emulates GenServer
 */
public class GenericServer {
    public static final OtpErlangAtom GEN_CALL = new OtpErlangAtom("$gen_call");

    public static OtpErlangObject call(OtpMbox callerMbox, OtpNode callerNode, String serverName, String serverNode, OtpErlangObject request, int timeout) throws OtpErlangExit, OtpErlangDecodeException {
        OtpErlangPid serverPid = Process.whereis(callerMbox, callerNode, serverName, serverNode, timeout);

        if (serverPid == null) {
            throw new OtpErlangExit("no process");
        }

        return call(callerMbox,  callerNode, serverPid, request, timeout);
    }

    private static OtpErlangObject call(@NotNull OtpMbox callerMbox, @NotNull OtpNode callerNode, @NotNull OtpErlangPid serverPid, @NotNull OtpErlangObject request, int timeout) throws OtpErlangExit, OtpErlangDecodeException {
        callerMbox.link(serverPid);

        OtpErlangObject response = unmonitoredCall(callerMbox, callerNode, serverPid, request, timeout);

        callerMbox.unlink(serverPid);

        return response;
    }

    private static OtpErlangObject unmonitoredCall(OtpMbox callerMbox, OtpNode callerNode, OtpErlangPid serverPid, OtpErlangObject request, int timeout) throws OtpErlangExit, OtpErlangDecodeException {
        return Generic.unmonitoredCall(
                callerMbox,
                callerNode,
                serverPid,
                GEN_CALL,
                request,
                timeout
        );
    }

    public static OtpErlangObject unmonitoredCall(OtpMbox callerMbox, OtpNode callerNode, String serverName, String serverNode, OtpErlangObject request, int timeout) throws OtpErlangExit, OtpErlangDecodeException {
        return Generic.unmonitoredCall(
                callerMbox,
                callerNode,
                serverName,
                serverNode,
                GEN_CALL,
                request,
                timeout
        );
    }
}

File: src/org/elixir_lang/GenericServer.kt
package org.elixir_lang

import com.ericsson.otp.erlang.*

/**
 * Emulates GenServer
 */
object GenericServer {
    @Throws(OtpErlangExit::class, OtpErlangDecodeException::class)
    fun call(
            callerMbox: OtpMbox,
            callerNode: OtpNode,
            serverName: String,
            serverNode: String,
            request: OtpErlangObject,
            timeout: Int
    ): OtpErlangObject? {
        val serverPid = Process.whereis(callerMbox, callerNode, serverName, serverNode, timeout)
                ?: throw OtpErlangExit("Could not determine PID for $serverName on $serverNode within $timeout")

        return call(callerMbox, callerNode, serverPid, request, timeout)
    }

    @Throws(OtpErlangExit::class, OtpErlangDecodeException::class)
    private fun call(
            callerMbox: OtpMbox,
            callerNode: OtpNode,
            serverPid: OtpErlangPid,
            request: OtpErlangObject,
            timeout: Int
    ): OtpErlangObject? {
        callerMbox.link(serverPid)

        val response = unmonitoredCall(callerMbox, callerNode, serverPid, request, timeout)

        callerMbox.unlink(serverPid)

        return response
    }

    @Throws(OtpErlangExit::class, OtpErlangDecodeException::class)
    private fun unmonitoredCall(
            callerMbox: OtpMbox,
            callerNode: OtpNode,
            serverPid: OtpErlangPid,
            request: OtpErlangObject,
            timeout: Int
    ): OtpErlangObject? =
            Generic.unmonitoredCall(
                    callerMbox,
                    callerNode,
                    serverPid,
                    GEN_CALL,
                    request,
                    timeout
            )

    @Throws(OtpErlangExit::class, OtpErlangDecodeException::class)
    fun unmonitoredCall(
            callerMbox: OtpMbox,
            callerNode: OtpNode,
            serverName: String,
            serverNode: String,
            request: OtpErlangObject,
            timeout: Int
    ): OtpErlangObject? =
            Generic.unmonitoredCall(
                    callerMbox,
                    callerNode,
                    serverName,
                    serverNode,
                    GEN_CALL,
                    request,
                    timeout
            )

    val GEN_CALL = OtpErlangAtom("\$gen_call")
    val GEN_CAST = OtpErlangAtom("\$gen_cast")
}
